osquery/tables/system/windows/users.cpp

```diff
@@ -65,6 +65,46 @@ std::string getUserHomeDir(const std::string& sid) {
   return "";
 }
 
+// Given a SID, retrieve information about the matching user
+void genUser(const std::string& sidString, QueryData& results) {
+  Row r;
+
+  r["uuid"] = sidString;
+  r["directory"] = getUserHomeDir(sidString);
+
+  PSID sid;
+  auto ret = ConvertStringSidToSidA(sidString.c_str(), &sid);
+  if (ret == 0) {
+    VLOG(1) << "Convert SID to string failed with " << GetLastError();
+  }
+  auto uid = getUidFromSid(sid);
+  auto gid = getGidFromSid(sid);
+  r["uid"] = BIGINT(uid);
+  r["gid"] = BIGINT(gid);
+  r["uid_signed"] = INTEGER(uid);
+  r["gid_signed"] = INTEGER(gid);
+  r["type"] = kWellKnownSids.find(sidString) == kWellKnownSids.end()
+                  ? "roaming"
+                  : "special";
+
+  // TODO
+  r["shell"] = "C:\\Windows\\system32\\cmd.exe";
```

Comment:
I know we're talking about this in github, maybe shove this into a function so it's not duplicated?

Reply:
You know what, good point. The code uses a level 3 field, `usri3_name`, to call `NetUserGetInfo()` with level 4. But it could have grabbed that field from a level 1 struct. Later, where it is generating row data using level 3 fields, it could be referencing the level 4 struct it already has from `NetUserGetInfo` instead. I'll make that change.

Reply:
Thanks @directionless , this improves the execution time by ~10% in testing.